lib/std/math/Interval.zig

```diff
@@ -0,0 +1,110 @@
+const std = @import("../std.zig");
+
+/// Create an interval of type T.
+/// See https://en.wikipedia.org/wiki/Interval_(mathematics) for more info.
+pub fn Interval(comptime T: type) type {
+    comptime std.debug.assert(std.meta.trait.isNumber(T)); // Interval must be instantiated with a number type.
+    return struct {
+        /// The left hand side of the Interval.
+        /// n as null here represents infinity.
+        /// open determines if the interval should be open (exclusive) or closed (inclusive). 
+        from: Endpoint,
+        /// The right hand side of the Interval.
+        /// n as null here represent infinity.
+        /// open determines if the interval should be open (exclusive) or closed (inclusive). 
+        to: Endpoint,
+
+        pub const Endpoint = union(enum) {
+            Infinity,
```

Comment:
this doesn't make sense for floating point types

Reply:
Unless you're working with bigints it doesn't even make sense for non-fp types, the number of bits is finite and the upper/lower limit is well known.

Reply:
why doesn't it make sense?

Reply:
You can just make a closed endpoint with infinity (fp) or the max/min possible value (int).

Reply:
perhaps but including `.infinity` makes intent clearer when constructing the value and removes a branch when in `.contains(x)`

Reply:
Just make a function that returns the infinity value.
I don't see how it would result in less branches. LLVM is definitely smart enough to optimize out `num <= math.maxInt(T)`.
In the general case there are more branches, because you have to test for `.infinity`.

Reply:
thanks for the insight everyone, contains has now been cleaned up significantly